### PR TITLE
Sc13 sex contam geno wip

### PR DESCRIPTION
**Pull request by sc13-bioinf**
Adds sex determination, angsd and gatk haplotype caller.

Need to decide on a home for:
bin/Sex.DetERRmine.py from https://github.com/TCLamnidis/Sex.DetERRmine
assets/angsd_GRCh37_HapMapX.gz from /projects1/tools/angsd_0.910/RES/HapMapChrX.gz

I guess I could just add them for now?

Contributes to #40

### CODE REVIEW

**Comment by apeltzer** [comment]
I'd more enjoy seeing these as bioconda recipes 👍

**Comment by apeltzer** [comment]
`HapMapChrX` can go to `assets`

**Comment by apeltzer** [comment]
https://github.com/bioconda/bioconda-recipes/pull/15022

File: main.nf [hunk 1]
```diff
@@ -373,6 +397,40 @@ if( params.readPaths ){
 
 }
 
+// Setup for Pop Gen
```
Comment (apeltzer) [inline main.nf]: This kind of set up is not the regular way to do things in Nextflow, as it hides quite some complexity that one could easily miss and then wonder whats going on. The typicaly way would be:

- adding some parameters e.g. turning on sections and then using the `when: params.foo` to turn on a module. Also using `channel.mix(...)` as a directive to get output passed along if there are Switches. 

As the Haplotypecaller isn't entirely done (missing some parts from EAGER1, e.g. the GATK best practice part is not entirely there), I would also remove this from this PR and only add one piece at a time in a PR.

File: main.nf [hunk 2]
```diff
@@ -220,6 +233,17 @@ params.bamutils_clip_left = 1
 params.bamutils_clip_right = 1 
 params.bamutils_softclip = false 
 
+//Pop Gen settings
+
+angsd_region_grch37 = "X:2699520-154931044"
```
Comment (apeltzer) [inline main.nf]: this should be a `params.angsd_region = ...` setting, ideally set to some defaults in the  `nextflow.config`.

File: main.nf [hunk 3]
```diff
@@ -153,6 +165,7 @@ params.pmd_udg_type = 'half'
 params.multiqc_config = "$baseDir/conf/multiqc_config.yaml"
 params.email = false
 params.plaintext_email = false
+params.pop_gen_bedfile = "$baseDir/1240K.pos.list_hs37d5.0based.bed"
```
Comment (apeltzer) [inline main.nf]: Can we bundle this for legal reasons with the pipeline? If yes, nice - if not, we should provide it via a parameter.

File: main.nf [hunk 4]
```diff
@@ -111,6 +115,9 @@ def helpMessage() {
       --bamutils_clip_left / --bamutils_clip_right  Specify the number of bases to clip off reads
       --bamutils_softclip           Use softclip instead of hard masking
 
+    Genotyping
+      --genotyping_haplo_caller     Genotype calling using GATK HaplotypeCaller
```
Comment (apeltzer) [inline main.nf]: I'd say just do Sex Determination + X Chromosomal contamination tests, lets do the Genotyping in a separate pull request once this is done to keep things cleaner 👍

File: main.nf [hunk 5]
```diff
@@ -37,6 +37,10 @@ def helpMessage() {
 
     Input Data Additional Options:
       --snpcapture                  Runs in SNPCapture mode (specify a BED file if you do this!)
+      --pop_gen                     Run sex determination and angsd contamination estimation (if male)
+      --pop_gen_genome_build        GRCh37 or GRCh38 (Defaults to genome{0,5} and is required to set the PAR region for contamination estimation when this is not explicitly supplied)
```
Comment (apeltzer) [inline main.nf]: Please not ;-) There is a `--genome` switch already (which I intend to use for everything, especially in human genetics who are always using the same reference genome) and we can do a simple `params.fasta.contains("GRCH37") | ...` in the process section for that specific part to check this easily. I would simply use that instead of adding another parameter that isn't strictly necessary 👍

File: main.nf [hunk 6]
```diff
@@ -1120,10 +1181,14 @@ ch_dedup_for_pmdtools = Channel.empty()
 //Bamutils TrimBam Channel
 ch_for_bamutils = Channel.empty()
 
+//Population genetics Channel
+ch_for_pop_gen = Channel.empty()
```
Comment (apeltzer) [inline main.nf]: I did that before as well, but we should try to not use such setups. If its input, then add it to a channel. If its coming from another process, create the channel in that specific process.

File: main.nf [hunk 7]
```diff
@@ -1190,8 +1255,80 @@ process bam_trim {
     """
 }
 
+ch_pop_gen = Channel.create()
```
Comment (apeltzer) [inline main.nf]: Don't create empty channels if its not strictly necessary.

File: main.nf [hunk 8]
```diff
@@ -1190,8 +1255,80 @@ process bam_trim {
...
+ch_bam_for_pop_gen = Channel.empty()
+
+if(!params.skip_deduplication){
```
Comment (apeltzer) [inline main.nf]: This kind of logic should stay in processes if possible.

File: main.nf [hunk 9]
```diff
@@ -1190,8 +1255,80 @@ process bam_trim {
...
+    ch_pop_gen.close()
+}
+
+
+process sex_determination {
+
+    when: params.pop_gen
+
+    input:
+    file(bam_file) from ch_pop_gen
```
Comment (apeltzer) [inline main.nf]: if it just takes a bamfile, you could simply use any bamfile from any upstream tool and mix the channels.

File: main.nf [hunk 10]
```diff
@@ -1190,8 +1255,80 @@ process bam_trim {
...
+    angsd -i ${bam_file} -r ${pop_gen_angsd_region} -doCounts 1 -iCounts 1 -minMapQ 30 -minQ 30 -out ${bam_file.baseName}.doCounts
+    contamination -a ${bam_file.baseName}.doCounts.icnts.gz -h ${pop_gen_angsd_poly_sites} 2> ${bam_file.baseName}.X.contamination.out
+    export ANGSD_CONT_ET=\$(cat ${bam_file.baseName}.X.contamination.out | grep "Method2: new_llh" | cut -d ' ' -f 4 | cut -d ':' -f 2)
+    export ANGSD_CONT_SE=\$(cat ${bam_file.baseName}.X.contamination.out | grep "Method2: new_llh" | cut -d ' ' -f 5 | cut -d ':' -f 2)
+    echo -e "${bam_file.baseName}\\t\${ANGSD_CONT_ET}\\t\${ANGSD_CONT_SE}" > ${bam_file.baseName}.X.contamination.out.tsv
+    cat ${bam_file.baseName}.X.contamination.out.tsv | cut -f 2
+    """
+}
+
+process gatk_haplotype_caller {
```
Comment (apeltzer) [inline main.nf]: Please do a separate PR for this. I'd like to have both UnifiedGenotyper and HaplotypeCaller implemented in one PR, making that "package" available in general. :-)

Comment (jfy133) [reply main.nf]: #232 although extra feedback for HC would be welcome as it looks like there are different options supplied here.

**Comment by apeltzer** [comment]
Closing this in favor of #246
